### PR TITLE
CORE-956 Added 'find' functionality to BlockStore. Moved InMemBlockStore to tests. Import cleanup in Casper.

### PR DESCRIPTION
**Pull request by dzajkowski**
## Overview
A `find by predicate` method added to BlockStore.

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://rchain.atlassian.net/projects/CORE/issues/CORE-956

### Complete this checklist before you submit the PR
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
none

### CODE REVIEW

File: block-storage/src/main/scala/coop/rchain/blockstorage/LMDBBlockStore.scala [hunk 1]
```diff
@@ -84,6 +84,25 @@ class LMDBBlockStore[F[_]] private (val env: Env[ByteBuffer], path: Path, blocks
             }
     } yield ret
 
+  override def find(p: BlockHash => Boolean): F[Seq[(BlockHash, BlockMessage)]] =
+    for {
+      _ <- metricsF.incrementCounter(MetricNamePrefix + "find")
+      ret <- withReadTxn { txn =>
+              blocks
+                .iterate(txn)
+                .iterable()
```
Comment (birchmd) [inline block-storage/src/main/scala/coop/rchain/blockstorage/LMDBBlockStore.scala]: Why turn into an `Iterable` here? I'm not too sure about the LMDB API overall, but I would wait until after the filter is applied.

File: block-storage/src/test/scala/coop/rchain/blockstorage/BlockStoreTest.scala [hunk 2]
```diff
@@ -106,6 +106,23 @@ trait BlockStoreTest
...
+      forAll(blockStoreElementsGen, minSize(0), sizeRange(10)) { blockStoreElements =>
+        val items = blockStoreElements
+        items.foreach(store.put(_))
+        items.foreach {
+          case (k, v) =>
+            val w = store.find(_ == ByteString.copyFrom(k.getBytes()))
+            w should have size 1
+            w.head._2 shouldBe v
+        }
+        store.asMap().size shouldEqual items.size
```
Comment (lukasz-golebiewski) [inline block-storage/src/test/scala/coop/rchain/blockstorage/BlockStoreTest.scala]: Should we use deprecated methods in tests?

File: block-storage/src/main/scala/coop/rchain/blockstorage/BlockStore.scala [hunk 3]
```diff
@@ -15,6 +15,8 @@ trait BlockStore[F[_]] {
 
   def get(blockHash: BlockHash): F[Option[BlockMessage]]
 
+  def find(p: BlockHash => Boolean): F[Seq[(BlockHash, BlockMessage)]]
```
Comment (lukasz-golebiewski) [inline block-storage/src/main/scala/coop/rchain/blockstorage/BlockStore.scala]: Is it possible for the result to contain a hash that was not already supplied as parameter? If not - why return it?

Comment (dzajkowski) [reply block-storage/src/main/scala/coop/rchain/blockstorage/BlockStore.scala]: I would say that it's deprecated for external use. after it's not needed I would make it package private and let it be for tests. or we can try to replace it with `find`. I'm fine with that also.

File: block-storage/src/test/scala/coop/rchain/blockstorage/InMemBlockStore.scala [hunk 4]
```diff
@@ -30,6 +30,12 @@ class InMemBlockStore[F[_], E] private ()(implicit
       state <- refF.get
     } yield state
 
+  override def find(p: BlockHash => Boolean): F[Seq[(BlockHash, BlockMessage)]] =
+    for {
+      _     <- metricsF.incrementCounter("block-store-find")
+      state <- refF.get
+    } yield state.toSeq.filter { case (k, _) => p(k) }
```
Comment (birchmd) [inline block-storage/src/test/scala/coop/rchain/blockstorage/InMemBlockStore.scala]: I suspect that `state.filterKeys(p).toSeq` would be a little more efficient.

**Pull request merged**